### PR TITLE
138020087 optimizer functional

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
Added optimizer functional test.

### CODE REVIEW

**Comment by bhuvnesh2703** [comment]
@oarap @d @khannaekta @dhanashreek89 @jpatel-pivotal @vraghavan78 pls review

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_bit.ans [hunk 1]
```diff
@@ -28,6 +28,8 @@ SELECT * FROM dml_bit ORDER BY 1;
 -- OUT OF RANGE VALUES
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_bit.ans]: This whole dml_boundary_bit test seem redundant with what we have in src/test/regress/sql/bit.sql. I would suggest removing it altogether.

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_cidr.ans [hunk 2]
```diff
@@ -49,6 +49,8 @@ SELECT * FROM dml_cidr ORDER BY 1;
 -- OUT OF RANGE
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_cidr.ans]: This whole dml_boundary_cidr test seem redundant with what we have in src/test/regress/sql/inet.sql. I would suggest removing it altogether.

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_date.ans [hunk 3]
```diff
@@ -63,6 +63,8 @@ SELECT * FROM dml_date ORDER BY 1;
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_date.ans]: Most of the stuff in this file seems redundant with the tests in src/test/regress/sql/date.sql. The boundary tests around 4713-01-01 BC might be marginally useful, I don't see that in date.sql. But please move that out of TINC.

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_doubleprecision.ans [hunk 4]
```diff
@@ -44,6 +44,8 @@ SELECT * FROM dml_dp ORDER BY 1;
 -- OUT OF RANGE
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_doubleprecision.ans]: All the test queries in dml_boundary_doubleprecision test seem redundant with those in src/test/regress/float8.sql. I would suggest just removing them altogether from TINC.

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_inet.ans [hunk 5]
```diff
@@ -59,6 +61,8 @@ SELECT * FROM dml_inet ORDER BY 1;
 
 INSERT INTO dml_inet VALUES('');
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_inet.ans]: src/test/regress/sql/inet.sql has good coverage of these cases. I recommend removing dml_boundary_inet from TINC altogether.

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_macaddr.ans [hunk 6]
```diff
@@ -28,6 +28,8 @@ SELECT * FROM dml_macaddr ORDER BY 1;
 --INVALID VALUES
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_macaddr.ans]: A 'macaddr' test was added to PostgreSQL in 8.4. I would suggest backporting that instead, or just live without a special macaddr test, until we merge with 8.4 (we don't do anything special with macaddr in GPDB). And remove dml_boundary_macaddr test from TINC altogether.

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_real.ans [hunk 7]
```diff
@@ -38,6 +38,8 @@ SELECT * FROM dml_real ORDER BY 1;
 -- OUT OF RANGE VALUES
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_real.ans]: We have good coverage of the float4/real type in src/test/regress/sql/float4.sql. I'd suggest just removing this dml_boundary_real test from TINC.

File: src/test/tinc/tincrepo/dml/functional/expected/joins_ao_pttab_insert_20.ans [hunk 8]
```diff
@@ -16,6 +16,8 @@ LINE 1: SELECT COUNT(*) FROM (SELECT ('a')::int, dml_ao_pt_r.b,10 FR...
                                       ^
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/joins_ao_pttab_insert_20.ans]: All the queries in this test file seem completely uninteresting. That error is caught very early on in the parse analysis or planning stage, there's no reason to believe that having deep inside a complicated query would catch any more bugs than a straightforward `SELECT 'a'::int` test. And we have coverage for that in src/test/regress/sql/int4.sql already.

Recommendation: Remove joins_ao_pttab_insert_20 test from TINC.

File: src/test/tinc/tincrepo/dml/functional/expected/joins_ao_regtab_insert_20.ans [hunk 9]
```diff
@@ -21,6 +21,8 @@ LINE 1: SELECT COUNT(*) FROM (SELECT ('a')::int, dml_ao_r.b,10 FROM ...
                                       ^
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/joins_ao_regtab_insert_20.ans]: Like joins_ao_pttab_insert_20, this is totally uninteresting. Recommendation: remove from TINC.

File: src/test/tinc/tincrepo/dml/functional/expected/joins_co_pttab_insert_20.ans [hunk 10]
```diff
@@ -21,6 +21,8 @@ LINE 1: SELECT COUNT(*) FROM (SELECT ('a')::int, dml_co_pt_r.b,10 FR...
                                       ^
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/joins_co_pttab_insert_20.ans]: Like  joins_ao_pttab_insert_20, this is uninteresting. Recommendation: Remove from TINC.

File: src/test/tinc/tincrepo/dml/functional/expected/joins_heap_pttab_update_19.ans [hunk 11]
```diff
@@ -18,6 +18,8 @@ SELECT SUM(b) FROM dml_heap_pt_r;
 
 UPDATE dml_heap_pt_r SET b = MAX(dml_heap_pt_s.b) FROM dml_heap_pt_s WHERE dml_heap_pt_r.b = dml_heap_pt_s.a;
```
Comment (hlinnaka) [inline src/test/tinc/tincrepo/dml/functional/expected/joins_heap_pttab_update_19.ans]: Please move this out of TINC. src/test/regress/sql/update_gp.sql seems like a good place.

**Comment by cjcjameson** [comment]
Just to clarify, you're merging from your fork but what's https://github.com/greenplum-db/gpdb/tree/138020087-Optimizer-Functional -- same or different?

**Comment by bhuvnesh2703** [comment]
Thanks @hlinnaka for the feedback. Incorporated it. Let me know for any other comments.
Also it seems that we may need to review the complete suite and remove the redundant test as there are around 3000+ tests in optimizer functional and even remove the same from 4.3 as well. Have created a story to track it.

However will like to push this PR so that we can get it plugged to the 5 pipeline. The duration to run the test is around 20mins and it will run in parallel to icw so will not increase the timeline of the run.

**Comment by bhuvnesh2703** [comment]
@cjcjameson thanks for pointing out, changed the ref to master

File: src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_date.ans (hunk 3 above)
Comment (khannaekta) [inline src/test/tinc/tincrepo/dml/functional/expected/dml_boundary_date.ans]: The purpose of this suite is mostly testing the optimizer.

The boundary assertion about the lower bound of Julian date is meaningful but it doesn't exercise the optimizer at all. We would suggest we remove that altogether.

Jesse and Ekta

**Comment by khannaekta** [comment]
ship it!

**Pull request merged**